### ISSUE TITLE
Batch support & Backprop 

### ISSUE THREAD

**Issue by mpariente** [JuanFMontesinos/torch_mir_eval#1]
First thanks a lot for the implementation, that's great !

Is there a plan to support batch computation? This would really be a game changer ! 

Regarding the backprop, the only op that prevents it is the toeplitz matrix in numpy right? 
I'm going to make a PR for the toeplitz matrix.

**Comment by JuanFMontesinos** [comment]
Hi,
Awesome contribution :smile:
Well let me just say that this was created like a lazy implementation, a seed, to push the community to port everything to GPU.

WRT your questions:
**Backprop issues:**
D and G are built using in-place operations.
https://github.com/JuanFMontesinos/torch_mir_eval/blob/27f6e94d1646cfcd79d38b1effbcbddff7a4a89a/torch_mir_eval/separation.py#L290-L298
https://github.com/JuanFMontesinos/torch_mir_eval/blob/27f6e94d1646cfcd79d38b1effbcbddff7a4a89a/torch_mir_eval/separation.py#L302-L306

D and G should be composed by stacking tensors in order to make it backpropagable. 

**Batched version:**
First of all a definition of batch is required.
Do you mean the parallel computation of all the sources or the parallel computation of independent evaluations?

For the former I don't think gain in speed will be really significant. 
For the latter  there are two technical issues 
The solvers are not batched
https://github.com/JuanFMontesinos/torch_mir_eval/blob/27f6e94d1646cfcd79d38b1effbcbddff7a4a89a/torch_mir_eval/separation.py#L317-L318
The convolutions depends on kernels which depends on the sources

Lastly, right know this works like a drop-n-replace library, you can do
`import torch_mir_eval as mir_eval` and the original format and functions and imports are respected but you need to pass tensors instead of arrays. It's easy to replace in any code with minor changes.

Let me know what do you think. 
Juan

**Comment by JuanFMontesinos** [comment]
Hi,
Awesome contribution :smile:
Well let me just say that this was created like a lazy implementation, a seed, to push the community to port everything to GPU.

WRT your questions:
**Backprop issues:**
D and G are built using in-place operations.
https://github.com/JuanFMontesinos/torch_mir_eval/blob/27f6e94d1646cfcd79d38b1effbcbddff7a4a89a/torch_mir_eval/separation.py#L290-L298
https://github.com/JuanFMontesinos/torch_mir_eval/blob/27f6e94d1646cfcd79d38b1effbcbddff7a4a89a/torch_mir_eval/separation.py#L302-L306

D and G should be composed by stacking tensors in order to make it backpropagable. 

**Batched version:**
First of all a definition of batch is required.
Do you mean the parallel computation of all the sources or the parallel computation of independent evaluations?

For the former I don't think gain in speed will be really significant. 
For the latter  there are two technical issues 
The solvers are not batched
https://github.com/JuanFMontesinos/torch_mir_eval/blob/27f6e94d1646cfcd79d38b1effbcbddff7a4a89a/torch_mir_eval/separation.py#L317-L318
The convolutions depends on kernels which depends on the sources

Lastly, right know this works like a drop-n-replace library, you can do
`import torch_mir_eval as mir_eval` and the original format and functions and imports are respected but you need to pass tensors instead of arrays. It's easy to replace in any code with minor changes.

Let me know what do you think. 
Juan

**Comment by mpariente** [comment]
Thanks for your answer.

> **Backprop issues:**
> D and G are built using in-place operations.
> 
> D and G should be composed by stacking tensors in order to make it backpropagable.
> 

Even constructed like that, it seems backpropagable without problem. I'm pretty sure that indexing doesn't stop gradient flow. We'll just integrate #2 into the code and see if it's backpropagable. 

> **Batched version:**
> First of all a definition of batch is required.
> Do you mean the parallel computation of all the sources or the parallel computation of independent evaluations?

I meant to compute the SDR for a batch of independent entries, so that we can speed things up. 

> For the latter there are two technical issues
> The solvers are not batched

`torch.solve` supports the batch dimension, `torch.lstsq` doesn't seem to but we can work around that I guess. 

> 
> The convolutions depends on kernels which depends on the sources
> 

You can pack the source dimension in the batch dimension and have independent convolution filters like we do here. 

https://github.com/mpariente/asteroid/blob/master/asteroid/filterbanks/enc_dec.py#L185-192

> Lastly, right know this works like a drop-n-replace library, you can do
> `import torch_mir_eval as mir_eval` and the original format and functions and imports are respected but you need to pass tensors instead of arrays. It's easy to replace in any code with minor changes.

This is really nice indeed. 
And it would be nice to keep supporting that, but it doesn't prevent us to support batched computation i.e if the input array is 2D, perform the usual computation, if 3D perform batched computation. 

Overall, a batched version for the GPU would great because it would be much faster, and it would enable using it as a loss (one backpropagable).

**Comment by JuanFMontesinos** [comment]
> Even constructed like that, it seems backpropagable without problem. I'm pretty sure that indexing doesn't stop gradient flow. We'll just integrate #2 into the code and see if it's backpropagable.  
Seems you are right. I think it used not to be allowed. 

Soo you convinced me :) 

> You can pack the source dimension in the batch dimension and have independent convolution filters like we do here.

https://github.com/mpariente/asteroid/blob/master/asteroid/filterbanks/enc_dec.py#L185-192

I don't understand your what are you proposing here. Filters are

`filters of shape (out_channels,in_channel,kW`
```
    for i in range(nsrc):
        sproj += torch.nn.functional.conv1d(reference_sources[i][:-flen + 1][None, None, ...],
                                            C[:, i].flip([0])[None, None, ...],
                                            padding=flen - 1)[0, 0]
```
Non-batched case
In the case in the snippet above for each `reference_source[i]` we have a different kernel `C[:,i]`

So if you have a torch tensor of the  shape
nsrc,1,N
if you have a filter `O,1,K`

Each `reference_source[i]` is being convolved with O different filters.
But those filters are exactly the same for each element in the batch.
However here we need nsrc filters but we do not need to apply those nsrc filters to each `reference_source[i]`

In a batched case we would have
B,nsrc,1,N
We can map them as
Bxnsrc,1,N
but we yet need Bxnsrc filters and to apply each filter to a single element of that batch

I don't see how to parallelize that op without a for loop.

**Comment by mpariente** [comment]
I'm sorry, I'm not following the whole explanation, but I'm pretty sure that we can parallelize the current _for_ loop in the code, and also in the case of batched inputs. 
I'll have a look when I have time and show an example if I succeed. 

But the main question remains, would you be ok to integrate a batched version?

**Comment by JuanFMontesinos** [comment]
Yeh sure.
As soon as the toeplitz matrix is merged can do it :)   
Dunno if you want to finish the integration in `separation.py` or to merge as is 

The explanation in short means that pytorch's conv kernels doesnt have batch dim. So I don't see how to apply a single filter per batch element without loops or increasing the amount of filters in the kernel. But if you can do so then the extension to batched metrics is trivial.

**Comment by mpariente** [comment]
You exactly increase the number of filters in the kernel. View the signals as `(batch * n_src, ...)` instead of `(batch, n_src, ...)`, have `batch * n_filters_per batch` filters (and maybe use the `group` argument, not sure) and that would be it.

**Comment by JuanFMontesinos** [comment]
Hi,
Everything done at 3d4694a47ffbabcebe977aef88ace35c2dc70635
Toeplitz function works like a charm.
Indeed backprop seems to be working too. 

The drawback of increasing the number of filters in the kernel is that you will be computing B*(B-1) unnecessary filters. 
So in short the input will be
nsrc* batch_size,1,N
and the kernel will be
nsrc* batch_size,1,K

the output tensor will be `nsrc* batch_size,nsrc* batch_size,N'`
but we only need
```
for i in range(nsrc*batch_size):
    needed_el=output[i,i,:]
```
It still can be faster than a for loop but prob not memory efficient. 
If nsrc*batch_size is high (~512/1024/2048) can lead to OOM issues.  

Did I understand u properly?

**Comment by JuanFMontesinos** [comment]
Done at 377fe51a6b08d43af63d6b7f7805578515713a92

**Comment by mpariente** [comment]
This is great !
